Implicit architecture targeting not in place for `dotnet tool install`

The spec talks about [implicit RID targeting](https://github.com/dotnet/designs/blob/main/accepted/2021/x64-emulation-on-arm64/x64-emulation.md#rely-on-implicit-rid-defaults).

From that doc, these are the implicit RIDs:

- .NET 6: osx-arm64
- .NET 5: osx-x64
- .NET Core 3.1: osx-x64

Not documented, on Windows these would be slightly different:

- .NET 6: win-arm64
- .NET 5: win-arm64
- .NET Core 3.1: win-x64

My environment:

```bash
rich@MacBook-Air-M1-2020 app % dotnet --info | grep RID
RID: osx-arm64
rich@MacBook-Air-M1-2020 app % dotnet --version
6.0.100-rc.2.21478.25
```

This is working, AFAICT, for `dotnet run`.

```bash
rich@MacBook-Air-M1-2020 app % cat app.csproj | grep Target
<TargetFramework>net5.0</TargetFramework>
rich@MacBook-Air-M1-2020 app % dotnet run
/Users/rich/app/app.csproj : error NU1102: Unable to find package Microsoft.NETCore.App.Host.osx-x64 with version (= 5.0.11)
/Users/rich/app/app.csproj : error NU1102: - Found 2021 version(s) in dotnet6 [ Nearest version: 6.0.0-alpha.1.20423.5 ]
/Users/rich/app/app.csproj : error NU1102: - Found 62 version(s) in nuget.org [ Nearest version: 6.0.0-preview.1.21102.12 ]
/Users/rich/app/app.csproj : error NU1102: - Found 1 version(s) in internal [ Nearest version: 6.0.0-rc.2.21474.18 ]

The build failed. Fix the build errors and run again.
```

That's failing, but due to a mismatch in versions available on the feeds I have registered, so NP. The key thing is that the CLI is asking for the `osx-x64` package, which is the desired one.

Let's look at `dotnet tool install` in contrast.

```bash   
rich@MacBook-Air-M1-2020 app % dotnet tool install -g dotnet-serve
You can invoke the tool using the following command: dotnet-serve
Tool 'dotnet-serve' (version '1.9.71') was successfully installed.
rich@MacBook-Air-M1-2020 app % dotnet-serve 
zsh: killed     dotnet-serve
rich@MacBook-Air-M1-2020 app % codesign -s - ~/.dotnet/tools/dotnet-serve 
rich@MacBook-Air-M1-2020 app % dotnet-serve                              
It was not possible to find any compatible framework version
The framework 'Microsoft.AspNetCore.App', version '5.0.0' (arm64) was not found.
  - The following frameworks were found:
      6.0.0-rc.2.21470.37 at [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
      6.0.0-rc.2.21475.12 at [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]

You can resolve the problem by installing the specified framework and/or SDK.

The specified framework can be found at:
  - https://aka.ms/dotnet-core-applaunch?framework=Microsoft.AspNetCore.App&framework_version=5.0.0&arch=arm64&rid=osx.11.1-arm64
rich@MacBook-Air-M1-2020 app % 
```

It is  pretty obvious that this  isn't a good situation.  The user needs to know a-priori that  `dotnet-serve` needs to be installed with `-a x64`. We  can and should do that automatically, matching `dotnet run` behavior.

--- a/src/Cli/dotnet/ShellShim/ShellShimTemplateFinder.cs
+++ b/src/Cli/dotnet/ShellShim/ShellShimTemplateFinder.cs
@@ -32,16 +32,16 @@ namespace Microsoft.DotNet.ShellShim
             _packageSourceLocation = packageSourceLocation;
         }
 
-        public async Task<string> ResolveAppHostSourceDirectoryAsync(string archOption, NuGetFramework targetFramework)
+        public async Task<string> ResolveAppHostSourceDirectoryAsync(string archOption, NuGetFramework targetFramework, Architecture arch)
         {
             string rid;
             var validRids = new string[] { "win-x64", "win-arm64", "osx-x64", "osx-arm64" };
             if (string.IsNullOrEmpty(archOption))
             {
                 if (targetFramework != null &&
-                    ((targetFramework.Version < new Version("6.0") && OperatingSystem.IsMacOS()) ||
-                    (targetFramework.Version < new Version("5.0") && OperatingSystem.IsWindows())
-                    && !RuntimeInformation.ProcessArchitecture.Equals(Architecture.X64)))
+                    ((targetFramework.Version.Major < 6 && OperatingSystem.IsMacOS()) ||
+                    (targetFramework.Version.Major < 5 && OperatingSystem.IsWindows())
+                    && !arch.Equals(Architecture.X64)))
                 {
                     rid = OperatingSystem.IsWindows() ? "win-x64" : "osx-x64";
                 }

--- a/src/Cli/dotnet/ShellShim/ShellShimTemplateFinder.cs
+++ b/src/Cli/dotnet/ShellShim/ShellShimTemplateFinder.cs
@@ -39,8 +39,8 @@ namespace Microsoft.DotNet.ShellShim
             if (string.IsNullOrEmpty(archOption))
             {
                 if (targetFramework != null &&
-                    ((targetFramework.Version.Major < 6 && OperatingSystem.IsMacOS()) ||
-                    (targetFramework.Version.Major < 5 && OperatingSystem.IsWindows())
+                    (((targetFramework.Version.Major < 6 && OperatingSystem.IsMacOS()) ||
+                    (targetFramework.Version.Major < 5 && OperatingSystem.IsWindows()))
                     && !arch.Equals(Architecture.X64)))
                 {
                     rid = OperatingSystem.IsWindows() ? "win-x64" : "osx-x64";

--- a/src/Cli/dotnet/ShellShim/ShellShimTemplateFinder.cs
+++ b/src/Cli/dotnet/ShellShim/ShellShimTemplateFinder.cs
@@ -32,14 +32,16 @@ namespace Microsoft.DotNet.ShellShim
             _packageSourceLocation = packageSourceLocation;
         }
 
-        public async Task<string> ResolveAppHostSourceDirectoryAsync(string archOption, string targetFramework)
+        public async Task<string> ResolveAppHostSourceDirectoryAsync(string archOption, NuGetFramework targetFramework)
         {
             string rid;
             var validRids = new string[] { "win-x64", "win-arm64", "osx-x64", "osx-arm64" };
             if (string.IsNullOrEmpty(archOption))
             {
-                if (!string.IsNullOrEmpty(targetFramework) && new NuGetFramework(targetFramework).Version < new Version("6.0")
-                    && (OperatingSystem.IsWindows() || OperatingSystem.IsMacOS()) && !RuntimeInformation.ProcessArchitecture.Equals(Architecture.X64))
+                if (targetFramework != null &&
+                    ((targetFramework.Version < new Version("6.0") && OperatingSystem.IsMacOS()) ||
+                    (targetFramework.Version < new Version("5.0") && OperatingSystem.IsWindows())
+                    && !RuntimeInformation.ProcessArchitecture.Equals(Architecture.X64)))
                 {
                     rid = OperatingSystem.IsWindows() ? "win-x64" : "osx-x64";
                 }

--- a/src/Cli/dotnet/ToolPackage/IToolPackage.cs
+++ b/src/Cli/dotnet/ToolPackage/IToolPackage.cs
@@ -4,6 +4,7 @@
 using System;
 using System.Collections.Generic;
 using Microsoft.Extensions.EnvironmentAbstractions;
+using NuGet.Frameworks;
 using NuGet.Versioning;
 
 namespace Microsoft.DotNet.ToolPackage
@@ -21,5 +22,7 @@ namespace Microsoft.DotNet.ToolPackage
         IEnumerable<string> Warnings { get; }
 
         IReadOnlyList<FilePath> PackagedShims { get; }
+
+        IEnumerable<NuGetFramework> Frameworks { get; }
     }
 }

--- a/src/Cli/dotnet/ToolPackage/ToolPackageInstance.cs
+++ b/src/Cli/dotnet/ToolPackage/ToolPackageInstance.cs
@@ -12,6 +12,7 @@ using NuGet.ProjectModel;
 using NuGet.Versioning;
 using Microsoft.DotNet.Cli.Utils;
 using System.Threading;
+using NuGet.Frameworks;
 
 namespace Microsoft.DotNet.ToolPackage
 {
@@ -53,6 +54,8 @@ namespace Microsoft.DotNet.ToolPackage
             }
         }
 
+        public IEnumerable<NuGetFramework> Frameworks { get; private set; }
+
         private const string AssetsFileName = "project.assets.json";
         private const string ToolSettingsFileName = "DotnetToolSettings.xml";
 
@@ -76,6 +79,9 @@ namespace Microsoft.DotNet.ToolPackage
             _lockFile =
                 new Lazy<LockFile>(
                     () => new LockFileFormat().Read(assetsJsonParentDirectory.WithFile(AssetsFileName).Value));
+            var toolsPackagePath = Path.Combine(PackageDirectory.Value, Id.ToString(), Version.ToNormalizedString(), "tools");
+            Frameworks = Directory.GetDirectories(toolsPackagePath)
+                .Select(path => NuGetFramework.ParseFolder(Path.GetFileName(path)));
         }
 
         private IReadOnlyList<RestoredCommand> GetCommands()

--- a/src/Cli/dotnet/commands/dotnet-tool/install/ToolInstallGlobalOrToolPathCommand.cs
+++ b/src/Cli/dotnet/commands/dotnet-tool/install/ToolInstallGlobalOrToolPathCommand.cs
@@ -135,9 +135,9 @@ namespace Microsoft.DotNet.Tools.Tool.Install
                         targetFramework: _framework, verbosity: _verbosity);
 
                     var framework = string.IsNullOrEmpty(_framework) && package.Frameworks.Count() > 0  ?
-                        package.Frameworks.MaxBy(f => f.Version) :
+                        package.Frameworks.Where(f => f.Version < new Version(Product.Version)).MaxBy(f => f.Version) :
                         NuGetFramework.Parse(_framework);
-                    string appHostSourceDirectory = _shellShimTemplateFinder.ResolveAppHostSourceDirectoryAsync(_architectureOption, framework).Result;
+                    string appHostSourceDirectory = _shellShimTemplateFinder.ResolveAppHostSourceDirectoryAsync(_architectureOption, framework, RuntimeInformation.ProcessArchitecture).Result;
                     IShellShimRepository shellShimRepository = _createShellShimRepository(appHostSourceDirectory, toolPath);
 
                     foreach (var command in package.Commands)

--- a/src/Cli/dotnet/commands/dotnet-tool/install/ToolInstallGlobalOrToolPathCommand.cs
+++ b/src/Cli/dotnet/commands/dotnet-tool/install/ToolInstallGlobalOrToolPathCommand.cs
@@ -105,11 +105,8 @@ namespace Microsoft.DotNet.Tools.Tool.Install
                 toolPath = new DirectoryPath(_toolPath);
             }
 
-            string appHostSourceDirectory = _shellShimTemplateFinder.ResolveAppHostSourceDirectoryAsync(_architectureOption, _framework).Result;
-
             (IToolPackageStore toolPackageStore, IToolPackageStoreQuery toolPackageStoreQuery, IToolPackageInstaller toolPackageInstaller) =
                 _createToolPackageStoresAndInstaller(toolPath, _forwardRestoreArguments);
-            IShellShimRepository shellShimRepository = _createShellShimRepository(appHostSourceDirectory, toolPath);
 
             // Prevent installation if any version of the package is installed
             if (toolPackageStoreQuery.EnumeratePackageVersions(_packageId).FirstOrDefault() != null)
@@ -136,6 +133,12 @@ namespace Microsoft.DotNet.Tools.Tool.Install
                         packageId: _packageId,
                         versionRange: versionRange,
                         targetFramework: _framework, verbosity: _verbosity);
+
+                    var framework = string.IsNullOrEmpty(_framework) && package.Frameworks.Count() > 0  ?
+                        package.Frameworks.MaxBy(f => f.Version) :
+                        NuGetFramework.Parse(_framework);
+                    string appHostSourceDirectory = _shellShimTemplateFinder.ResolveAppHostSourceDirectoryAsync(_architectureOption, framework).Result;
+                    IShellShimRepository shellShimRepository = _createShellShimRepository(appHostSourceDirectory, toolPath);
 
                     foreach (var command in package.Commands)
                     {

--- a/src/Tests/Microsoft.DotNet.PackageInstall.Tests/ToolPackageInstallerTests.cs
+++ b/src/Tests/Microsoft.DotNet.PackageInstall.Tests/ToolPackageInstallerTests.cs
@@ -29,6 +29,7 @@ using Microsoft.NET.TestFramework.Assertions;
 using Microsoft.NET.TestFramework.Commands;
 using Xunit.Abstractions;
 using Microsoft.NET.TestFramework.Utilities;
+using NuGet.Frameworks;
 
 namespace Microsoft.DotNet.PackageInstall.Tests
 {
@@ -844,6 +845,7 @@ namespace Microsoft.DotNet.PackageInstall.Tests
             package.Id.Should().Be(TestPackageId);
             package.Version.ToNormalizedString().Should().Be(TestPackageVersion);
             package.PackageDirectory.Value.Should().Contain(store.Root.Value);
+            package.Frameworks.Should().BeEquivalentTo(TestFrameworks);
 
             storeQuery.EnumeratePackageVersions(TestPackageId)
                 .Select(p => p.Version.ToNormalizedString())
@@ -976,8 +978,10 @@ namespace Microsoft.DotNet.PackageInstall.Tests
             if (useMock)
             {
                 fileSystem = new FileSystemMockBuilder().Build();
+                var frameworksMap = new Dictionary<PackageId, IEnumerable<NuGetFramework>>()
+                        { {TestPackageId, TestFrameworks } };
                 WriteNugetConfigFileToPointToTheFeed(fileSystem, writeLocalFeedToNugetConfig);
-                var toolPackageStoreMock = new ToolPackageStoreMock(root, fileSystem);
+                var toolPackageStoreMock = new ToolPackageStoreMock(root, fileSystem, frameworksMap);
                 store = toolPackageStoreMock;
                 storeQuery = toolPackageStoreMock;
                 installer = new ToolPackageInstallerMock(
@@ -988,7 +992,8 @@ namespace Microsoft.DotNet.PackageInstall.Tests
                         reporter: reporter,
                         feeds: feeds == null
                             ? GetMockFeedsForConfigFile(writeLocalFeedToNugetConfig)
-                            : feeds.Concat(GetMockFeedsForConfigFile(writeLocalFeedToNugetConfig)).ToList()));
+                            : feeds.Concat(GetMockFeedsForConfigFile(writeLocalFeedToNugetConfig)).ToList()),
+                    frameworksMap: frameworksMap);
                 uninstaller = new ToolPackageUninstallerMock(fileSystem, toolPackageStoreMock);
             }
             else
@@ -1054,6 +1059,7 @@ namespace Microsoft.DotNet.PackageInstall.Tests
         private readonly string _testTargetframework = BundledTargetFramework.GetTargetFrameworkMoniker();
         private const string TestPackageVersion = "1.0.4";
         private static readonly PackageId TestPackageId = new PackageId("global.tool.console.demo");
+        private static readonly IEnumerable<NuGetFramework> TestFrameworks = new NuGetFramework[] { NuGetFramework.Parse("netcoreapp2.1")};
 
         public ToolPackageInstallerTests(ITestOutputHelper log) : base(log)
         {

--- a/src/Tests/Microsoft.DotNet.ShellShim.Tests/ShellShimRepositoryTests.cs
+++ b/src/Tests/Microsoft.DotNet.ShellShim.Tests/ShellShimRepositoryTests.cs
@@ -22,6 +22,8 @@ using Microsoft.NET.TestFramework.Assertions;
 using Microsoft.NET.TestFramework.Commands;
 using Xunit;
 using Xunit.Abstractions;
+using NuGet.Frameworks;
+using Microsoft.DotNet.Cli.NuGetPackageDownloader;
 
 namespace Microsoft.DotNet.ShellShim.Tests
 {
@@ -401,6 +403,17 @@ namespace Microsoft.DotNet.ShellShim.Tests
                     string.Format(
                            CommonLocalizableStrings.MoreThanOnePackagedShimAvailable,
                            string.Join(';', filePaths)));
+        }
+
+        [WindowsOnlyTheory]
+        [InlineData("net5.0")]
+        [InlineData("netcoreapp3.1")]
+        public void WhenRidNotSupportedOnWindowsItIsImplicit(string tfm)
+        {
+            var tempDir = _testAssetsManager.CreateTestDirectory(identifier: tfm).Path;
+            var templateFinder = new ShellShimTemplateFinder(new MockNuGetPackageDownloader(), new DirectoryPath(tempDir), null);
+            var path = templateFinder.ResolveAppHostSourceDirectoryAsync(null, NuGetFramework.Parse(tfm), Architecture.Arm64).Result;
+            path.Should().Contain(tfm.Equals("net5.0") ? $"win-{RuntimeInformation.ProcessArchitecture.ToString()}" : "AppHostTemplate");
         }
 
         private static void MakeNameConflictingCommand(string pathToPlaceShim, string shellCommandName)

--- a/src/Tests/Microsoft.DotNet.ShellShim.Tests/ShellShimRepositoryTests.cs
+++ b/src/Tests/Microsoft.DotNet.ShellShim.Tests/ShellShimRepositoryTests.cs
@@ -413,7 +413,7 @@ namespace Microsoft.DotNet.ShellShim.Tests
             var tempDir = _testAssetsManager.CreateTestDirectory(identifier: tfm).Path;
             var templateFinder = new ShellShimTemplateFinder(new MockNuGetPackageDownloader(), new DirectoryPath(tempDir), null);
             var path = templateFinder.ResolveAppHostSourceDirectoryAsync(null, NuGetFramework.Parse(tfm), Architecture.Arm64).Result;
-            path.Should().Contain(tfm.Equals("net5.0") ? $"win-{RuntimeInformation.ProcessArchitecture.ToString()}" : "AppHostTemplate");
+            path.Should().Contain(tfm.Equals("net5.0") ? "AppHostTemplate" : "win-x64");
         }
 
         private static void MakeNameConflictingCommand(string pathToPlaceShim, string shellCommandName)

--- a/src/Tests/Microsoft.DotNet.Tools.Tests.ComponentMocks/ToolPackageInstallerMock.cs
+++ b/src/Tests/Microsoft.DotNet.Tools.Tests.ComponentMocks/ToolPackageInstallerMock.cs
@@ -11,6 +11,7 @@ using Microsoft.DotNet.Cli.Utils;
 using Microsoft.DotNet.ToolPackage;
 using Microsoft.Extensions.EnvironmentAbstractions;
 using Microsoft.NET.TestFramework.Utilities;
+using NuGet.Frameworks;
 using NuGet.Versioning;
 
 namespace Microsoft.DotNet.Tools.Tests.ComponentMocks
@@ -25,6 +26,7 @@ namespace Microsoft.DotNet.Tools.Tests.ComponentMocks
         private readonly Action _installCallback;
         private readonly Dictionary<PackageId, IEnumerable<string>> _warningsMap;
         private readonly Dictionary<PackageId, IReadOnlyList<FilePath>> _packagedShimsMap;
+        private readonly Dictionary<PackageId, IEnumerable<NuGetFramework>> _frameworksMap;
 
         public ToolPackageInstallerMock(
             IFileSystem fileSystem,
@@ -32,7 +34,8 @@ namespace Microsoft.DotNet.Tools.Tests.ComponentMocks
             ProjectRestorerMock projectRestorer,
             Action installCallback = null,
             Dictionary<PackageId, IEnumerable<string>> warningsMap = null,
-            Dictionary<PackageId, IReadOnlyList<FilePath>> packagedShimsMap = null)
+            Dictionary<PackageId, IReadOnlyList<FilePath>> packagedShimsMap = null,
+            Dictionary<PackageId, IEnumerable<NuGetFramework>> frameworksMap = null)
         {
             _fileSystem = fileSystem ?? throw new ArgumentNullException(nameof(fileSystem));
             _store = store ?? throw new ArgumentNullException(nameof(store));
@@ -40,6 +43,7 @@ namespace Microsoft.DotNet.Tools.Tests.ComponentMocks
             _installCallback = installCallback;
             _warningsMap = warningsMap ?? new Dictionary<PackageId, IEnumerable<string>>();
             _packagedShimsMap = packagedShimsMap ?? new Dictionary<PackageId, IReadOnlyList<FilePath>>();
+            _frameworksMap = frameworksMap ?? new Dictionary<PackageId, IEnumerable<NuGetFramework>>();
         }
 
         public IToolPackage InstallPackage(PackageLocation packageLocation, PackageId packageId,
@@ -96,8 +100,11 @@ namespace Microsoft.DotNet.Tools.Tests.ComponentMocks
                     IReadOnlyList<FilePath> packedShims = null;
                     _packagedShimsMap.TryGetValue(packageId, out packedShims);
 
+                    IEnumerable<NuGetFramework> frameworks = null;
+                    _frameworksMap.TryGetValue(packageId, out frameworks);
+
                     return new ToolPackageMock(_fileSystem, packageId, version,
-                        packageDirectory, warnings: warnings, packagedShims: packedShims);
+                        packageDirectory, warnings: warnings, packagedShims: packedShims, frameworks: frameworks);
                 },
                 rollback: () =>
                 {
@@ -156,6 +163,8 @@ namespace Microsoft.DotNet.Tools.Tests.ComponentMocks
             public IEnumerable<string> Warnings { get; set; }
 
             public IReadOnlyList<FilePath> PackagedShims { get; set; }
+
+            public IEnumerable<NuGetFramework> Frameworks => throw new NotImplementedException();
         }
     }
 }

--- a/src/Tests/Microsoft.DotNet.Tools.Tests.ComponentMocks/ToolPackageMock.cs
+++ b/src/Tests/Microsoft.DotNet.Tools.Tests.ComponentMocks/ToolPackageMock.cs
@@ -10,7 +10,7 @@ using Microsoft.DotNet.Cli;
 using Microsoft.DotNet.Cli.Utils;
 using Microsoft.DotNet.ToolPackage;
 using Microsoft.Extensions.EnvironmentAbstractions;
-
+using NuGet.Frameworks;
 using NuGet.Versioning;
 
 namespace Microsoft.DotNet.Tools.Tests.ComponentMocks
@@ -28,7 +28,8 @@ namespace Microsoft.DotNet.Tools.Tests.ComponentMocks
             NuGetVersion version,
             DirectoryPath packageDirectory,
             IEnumerable<string> warnings = null,
-            IReadOnlyList<FilePath> packagedShims = null)
+            IReadOnlyList<FilePath> packagedShims = null,
+            IEnumerable<NuGetFramework> frameworks = null)
         {
             _fileSystem = fileSystem ?? throw new ArgumentNullException(nameof(fileSystem));
             Id = id;
@@ -37,6 +38,7 @@ namespace Microsoft.DotNet.Tools.Tests.ComponentMocks
             _commands = new Lazy<IReadOnlyList<RestoredCommand>>(GetCommands);
             _warnings = warnings ?? new List<string>();
             _packagedShims = packagedShims ?? new List<FilePath>();
+            Frameworks = frameworks ?? new List<NuGetFramework>();
         }
 
         public PackageId Id { get; private set; }
@@ -61,6 +63,8 @@ namespace Microsoft.DotNet.Tools.Tests.ComponentMocks
                 return _packagedShims;
             }
         }
+
+        public IEnumerable<NuGetFramework> Frameworks { get; private set; }
 
         private IReadOnlyList<RestoredCommand> GetCommands()
         {

--- a/src/Tests/Microsoft.DotNet.Tools.Tests.ComponentMocks/ToolPackageStoreMock.cs
+++ b/src/Tests/Microsoft.DotNet.Tools.Tests.ComponentMocks/ToolPackageStoreMock.cs
@@ -8,6 +8,7 @@ using System.Linq;
 using Microsoft.DotNet.Cli;
 using Microsoft.DotNet.ToolPackage;
 using Microsoft.Extensions.EnvironmentAbstractions;
+using NuGet.Frameworks;
 using NuGet.Versioning;
 
 namespace Microsoft.DotNet.Tools.Tests.ComponentMocks
@@ -15,13 +16,16 @@ namespace Microsoft.DotNet.Tools.Tests.ComponentMocks
     internal class ToolPackageStoreMock : IToolPackageStoreQuery, IToolPackageStore
     {
         private IFileSystem _fileSystem;
+        private readonly Dictionary<PackageId, IEnumerable<NuGetFramework>> _frameworksMap;
 
         public ToolPackageStoreMock(
             DirectoryPath root,
-            IFileSystem fileSystem)
+            IFileSystem fileSystem,
+            Dictionary<PackageId, IEnumerable<NuGetFramework>> frameworksMap = null)
         {
             Root = new DirectoryPath(Path.GetFullPath(root.Value));
             _fileSystem = fileSystem ?? throw new ArgumentNullException(nameof(fileSystem));
+            _frameworksMap = frameworksMap ?? new Dictionary<PackageId, IEnumerable<NuGetFramework>>();
         }
 
         public DirectoryPath Root { get; private set; }
@@ -93,11 +97,15 @@ namespace Microsoft.DotNet.Tools.Tests.ComponentMocks
 
             foreach (var subdirectory in _fileSystem.Directory.EnumerateFileSystemEntries(packageRootDirectory.Value))
             {
+
+                IEnumerable<NuGetFramework> frameworks = null;
+                _frameworksMap.TryGetValue(packageId, out frameworks);
                 yield return new ToolPackageMock(
                     _fileSystem,
                     packageId,
                     NuGetVersion.Parse(Path.GetFileName(subdirectory)),
-                    new DirectoryPath(subdirectory));
+                    new DirectoryPath(subdirectory),
+                    frameworks: frameworks);
             }
         }
 

